proto/pluginapi/v1/dataspec.proto

```diff
@@ -21,7 +21,12 @@ message AttrSpec {
     CtyValue default_val = 3;
     CtyValue example_val = 4;
     string   doc         = 5;
-    bool     required    = 6;
+    // bool     required    = 6; // made obsolete by constraints system
+    uint32     constraints    = 7;
```

Comment:
Having constraints as bitmask in grpc creates a risk of accidentally breaking compatibility between different fabric cli & plugin versions if we modify or reorder them in Go code. To avoid this we should:
- Add tests in Go code to lock constraint values.
- Or use the list of constraint enum values in protobuf and encode/decode it into bitmask used in Go code.

As for constraints, I wonder if it's even a good idea to have them as a bitmask/enum type. Min, Max, OneOf could also be a constraints instead of separate properties if constraint was struct/message type.

Reply:
I've added the test. I've decided to use bitfield rather than other approaches, because there are potentially many constraints (and having tonnes of bools would be polluting the schema) and some combinations are more useful than others (i.e. `NonNull` and `Required`, because `Required` on its own accepts null values and we rarely want that). 

Bitfields seemed to me like the easiest way to communicate those sets of booleans (I expect that most code would use `RequiredMeaningfull`, but _some_ code would need something else, i.e. CSV delimiter can be a tab character, which would be trimmed as whitespace if `RequiredMeaningfull` is used)

> Min, Max, OneOf could also be a constraints

Yes, I've experimented with that, but there is really no other way to use (most) constraints except as properties of the attribute, and repeatedly creating a nested constraints struct would just increase the indentation level and character count without doing anything meaningful.